src/network/netdev/hsr.c

```diff
@@ -0,0 +1,142 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <net/if.h>
+#include <netinet/in.h>
+#include <linux/if_arp.h>
+
+#include "hsr.h"
+#include "string-table.h"
+
+static const char* const hsr_protocol_table[_NETDEV_HSR_PROTOCOL_MAX] = {
+        [NETDEV_HSR_PROTOCOL_HSR] = "HSR",
+        [NETDEV_HSR_PROTOCOL_PRP] = "PRP",
+};
+
+DEFINE_STRING_TABLE_LOOKUP(hsr_protocol, HsrProtocol);
+DEFINE_CONFIG_PARSE_ENUM(config_parse_hsr_protocol, hsr_protocol, HsrProtocol,
+                         "Failed to parse Protocol=");
+
+static int netdev_hsr_get_iface_indexes(Hsr *hsr, int *indexes) {
+        Link *link = NULL;
+        int r, i;
+
+        assert(hsr);
+
+        for (i = 0; i < _NETDEV_HSR_SLAVE_MAX; i++) {
+                r = link_get_by_name(hsr->meta.manager, hsr->slave_ifaces[i], &link);
+                if (r < 0)
+                        return r;
+
+                if (indexes)
+                        indexes[i] = link->ifindex;
+        }
+
+        return 0;
+}
+
+static int netdev_hsr_fill_message_create(NetDev *netdev, Link *link, sd_netlink_message *m) {
+        Hsr *hsr;
```

Comment:
Not a matter, but nowadays, we have `ASSERT_PTR()` macro. Hence you can declare the variable with assertions below like the following:
```c
Hsr *hsr = ASSERT_PTR(HSR(ASSERT_PTR(netdev)));
```